spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/StreamDefinition.java

```diff
@@ -127,25 +127,21 @@ public String toString() {
 
 
 	/**
-	 * Iterator that prevents mutation of its backing data structure.
-	 *
-	 * @param <T> the type of elements returned by this iterator
+	 * Iterator that walks through ModuleDefinitions in deployment order, that is, from downstream (sink, if present)
+	 * to upstream (source) module.
+	 * Also prevents mutation of its backing data structure.
 	 */
-	private static class ReadOnlyIterator<T> implements Iterator<T> {
-		private final Iterator<T> wrapped;
-
-		public ReadOnlyIterator(Iterator<T> wrapped) {
-			this.wrapped = wrapped;
-		}
+	public class DeploymentOrderIterator implements Iterator<ModuleDefinition> {
```

Comment:
In general I agree with passing along whatever state a new instance of a class needs into its constructor. Static inner classes also help avoid exposure of any of the parent class that isn't required.

However in this circumstance, the list is immutable and the iterator prevents mutation - so I’m OK with this non-static inner class having direct access to the list of modules. There is indeed a tighter coupling between the iterator and its parent stream definition, but I think that is appropriate here.